src/db/transformer.ts

```diff
@@ -0,0 +1,291 @@
+import {assert} from '../asserts';
+import {assertBTreeNode} from '../btree/node';
+import * as btree from '../btree/mod';
+import {
+  IndexChangeMeta,
+  Meta,
+  MetaTyped,
+  SnapshotMeta,
+  getRefs as getRefsFromCommitData,
+  assertCommitData,
+} from './commit';
+import type {Hash} from '../hash';
+import type * as db from './mod';
+import type * as dag from '../dag/mod';
+import type {ReadonlyJSONValue} from '../json';
+import {HashType} from './hash-type';
+import type {Value} from '../kv/store';
+
+export class Transformer {
```

Comment:
Could this be implemented on top of `db.Visitor`, it is also essentially a semantic visitor right?

Reply:
There are a lot of similarities. The Visitor class does not mutate the tree. The transformer may. So, we could get rid of the Visitor and only have the Transformer. The transformer is a lot more complicated though so I'm not sure we want to get rid of the Visitor.

One benefit of only having the Transformer is that it is one less thing to update/understand etc.

Reply:
We can revisit when things settle down (after the all the passes are implemented)